include/circt/Dialect/Handshake/HandshakeOps.td

```diff
@@ -1249,3 +1249,131 @@ def EndOp : Handshake_Op<"end", [Terminator]> {
     mlir::ValueRange getMemoryControls();
   }];
 }
+
+//===----------------------------------------------------------------------===//
+// Speculation operations
+//===----------------------------------------------------------------------===//
+
+def SpeculatorOp : Handshake_Op<"speculator", [ 
+  TypesMatchWith<"data operand type matches data result type",
+                 "dataIn", "dataOut", "$_self">]> {
+  let summary = "Central control unit of the speculative circuit.";
+  let description = [{
+    The speculator produces speculative tokens when real data has
+    not yet arrived. The speculator is also responsible for deciding 
+    the correctness of its speculation.
+
+    It needs to send control tokens to save units, commit units, 
+    and save-commit units.
+
+    Example:
+    ```mlir
+    %dataOut, %saveCtrl, %commitCtrl, %SCControl1, %SCControl2, 
+      %SCBranchControl = "handshake.speculator"(%dataOut) 
+      {name = #handshake.name<"speculator0">} : 
+      (i1) -> (i1, i1, i1, i3, i3, i1)
+    ```
+  }];
+
+  let arguments = (ins AnyType : $dataIn);
+  let results = (outs AnyType : $dataOut,
+                  I1 : $saveCtrl, I1 : $commitCtrl, 
+                  I<3> : $SCControl1, I<3> : $SCControl2, 
+                  I1 : $SCBranchControl);
+}
+
+def SpecSaveOp : Handshake_Op<"spec_save", [
+  TypesMatchWith<"data operand type matches data result type",
+                 "dataIn", "dataOut", "$_self">] > {
+  let summary = "Saves data tokens that interact in the speculative region.";
+  let description = [{
+    Save units are used to mark the beginning of the speculation region. 
+    Whenever a speculative token can interact with a non-speculative token, 
+    the non-speculative token needs to be saved. In case of correct speculation
+    or no speculation, the saved token can be dropped.
+
+    On the other hand, if the speculation was incorrect, the saved tokens 
+    need to be reinserted into the circuit to repeat the previously 
+    miscalculated computations.
+
+    Example:
+    ```mlir
+    %23 = "handshake.spec_save"(%22, %saveCtrl) 
```

Comment:
nit: use more descriptive SSA value names than `%22` and `%23` in examples. Same applies to other operation examples.